worker/upgradedatabase/worker.go

```diff
@@ -65,24 +89,47 @@ func (c Config) Validate() error {
 	return nil
 }
 
+type upgradeDBWorker struct {
+	catacomb catacomb.Catacomb
+
+	cfg Config
+
+	fromVersion version.Number
+	toVersion   version.Number
+}
+
 // NewUpgradeDatabaseWorker returns a new Worker.
-func NewUpgradeDatabaseWorker(config Config) (*Worker, error) {
+func NewUpgradeDatabaseWorker(config Config) (worker.Worker, error) {
 	if err := config.Validate(); err != nil {
 		return nil, errors.Trace(err)
 	}
 
-	w := &Worker{
-		upgradeComplete: config.UpgradeComplete,
-		agent:           config.Agent,
-		logger:          config.Logger,
+	w := &upgradeDBWorker{
+		cfg: config,
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
 	}
 
-	w.tomb.Go(w.loop)
 	return w, nil
 }
 
+// Kill implements worker.Worker.Kill.
```

Comment:
I still think we should do with workers what we did with the watcher base and embed a catacomb/tomb worker that hoists these. Different patch though.

Reply:
I have started doing this in another patch elsewhere. The problematic thing to consider is the loop/run part. It requires you to know upfront where the loop will be.